pennylane/devices/experimental/python_device.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file defines a prototype python device conforming to the new device interface.
+"""
+from typing import Sequence, Tuple, Union, Callable
+import multiprocessing
+
+from pennylane.tape import QuantumScript
+from pennylane.workflow import ExecutionConfig
+
+from .device_interface import AbstractDevice
+from .simulator import PlainNumpySimulator, adjoint_diff_gradient
+from .python_preprocessor import simple_preprocessor
+
+
+def _multiprocessing_single_execution(
+    task_i: int, single_qs: QuantumScript, return_dict: "multiprocessing.managers.DictProxy"
+):
+    """Perform a single execution. Wrapped for use with the ``multiprocessing`` module.
+
+    Args:
+        task_i (int): an integer corresponding to the place of the single qauntum script in a batch.
+        single_qs (QuantumScript): A single quantum script to execute
+        return_dict (multiprocessing.managers.DictProxy): A dictionary to store the result in.
+    """
+    simulator = PlainNumpySimulator()
+    return_dict[task_i] = simulator.execute(single_qs)
+
+
+class PythonDevice(AbstractDevice):
+    """Device containing a Python simulator favouring composition-like interface.
+
+    Keyword Args:
+        use_multiprocessing=False (Bool): whether or not to perform the execution of batches with ``multiprocessing``.
+
+    """
+
+    def __init__(self, use_multiprocessing=False):
+        self.use_multiprocessing = use_multiprocessing
+        super().__init__()
+
+    def execute(
+        self, qscripts: Union[QuantumScript, Sequence[QuantumScript]], execution_config=None
+    ):
+        if isinstance(qscripts, QuantumScript):
+            simulator = PlainNumpySimulator()
+            results = simulator.execute(qscripts)
```

Comment:
Is it a batch execute? In the simulator I only see a function for single execution.

Reply:
We deemed it unnecessary to have two functions for execution, so the public facing API should only be `execute` for all devices. A device should only be required to expose support for a QScript or an iterable thereof. If we want to handle this differently on a device by device level, it can also use a preprocessing step to iterate over, and instantiate.

In this way, we can handle both with the same API, and the device can either for-loop them, throw them to MPI/Dask/Ray/ProcessPoolExecutor, or whatever is preferable.  Namely, it gives more flexibility, the same feature set, and less code to support/test.

Do you think this is a worry?

Reply:
No I think it is a good idea, to have a uniform API to execute. I was looking for the reason `qml.execute` is outputing list of tuple for a list of single measurement quantum tape. See comment above

Reply:
Ah, gotcha; missed above, sorry!